# Исследование надежности заемщиков коммерческого банка

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления" data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Удаление дубликатов</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма" data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма</a></span></li><li><span><a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация дохода</a></span></li><li><span><a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация целей кредита</a></span></li><li><span><a href="#Поиск-явных-дубликатов,-итоговая-очистка-датафрейма" data-toc-modified-id="Поиск-явных-дубликатов,-итоговая-очистка-датафрейма-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Поиск явных дубликатов, итоговая очистка датафрейма</a></span></li><li><span><a href="#Влияние-различных-клиентских-характеристик-на-возврат-кредита-в-срок" data-toc-modified-id="Влияние-различных-клиентских-характеристик-на-возврат-кредита-в-срок-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Влияние различных клиентских характеристик на возврат кредита в срок</a></span></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Обзор данных

In [1]:
# импорт библиотек
import pandas as pd

In [2]:
# загрузка датасета (полный путь не приводится, т.к. датасет не размещен в открытом доступе)
df = pd.read_csv(path)

In [3]:
# получение первых 15 строк таблицы
df.head(15)

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,-8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875.639453,покупка жилья
1,1,-4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080.014102,приобретение автомобиля
2,0,-5623.422610,33,Среднее,1,женат / замужем,0,M,сотрудник,0,145885.952297,покупка жилья
3,3,-4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628.550329,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616.077870,сыграть свадьбу
5,0,-926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763.565419,покупка жилья
6,0,-2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525.971920,операции с жильем
7,0,-152.779569,50,СРЕДНЕЕ,1,женат / замужем,0,M,сотрудник,0,135823.934197,образование
8,2,-6929.865299,35,ВЫСШЕЕ,0,гражданский брак,1,F,сотрудник,0,95856.832424,на проведение свадьбы
9,0,-2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425.938277,покупка жилья для семьи


In [4]:
# получение общей информации о данных в таблице
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
# получение информации о названиях колонок (помогает визуально поискать скрытые пробелы)
df.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Таблица состоит из 21525 строк и 12 столбцов с данными следующих типов:
* вещественное число - 2 столбца;
* целое число - 5 столбцов;
* строка - 5 столбцов (могут содержаться данные смешанных типов).

Согласно документации к данным, столбцы таблицы содержат следующую информацию:
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

В названиях колонок нарушения стиля не обнаружено:
* все наименования записаны строчными буквами;
* пробелы отсутствуют;
* к названиям, состоящим из двух слов, применен "змеиный_регистр".

В столбцах `days_employed` и `total_income` есть пропущенные значения.

**Выводы**

В каждой строке таблицы — данные о заемщиках банка. Присутствуют колонки с описанием самого клиента и его социального статуса: пол, возраст, образование, семейное положение, количество детей, тип занятости, стаж. Есть данные о доходе клиента, а также информация о цели получения кредита и фактах нарушения графика платежей.

Предварительно можно утверждать, что данных достаточно для проверки гипотез о влиянии разных факторов на факт погашения кредита в срок.

Исследуем качество данных в исходном датасете и устраним проблемы в них.

## Заполнение пропусков

В столбцах `days_employed` и `total_income` есть пропущенные значения. В датасете они отражены как NaN (можно проводить математические операции), поскольку тип данных в соответствующих столбцах распознается как 'float' (иначе был тип 'object').

Найдем количество пропусков и их долю в общем количестве значений:

In [6]:
# .isna() преобразует столбец в Series из True (1 - пропуск есть) и False (0 - пропусков нет)
# .sum() посчитает сумму значений, т.е. по сути кол-во пропусков True
days_employed_null_number = df['days_employed'].isna().sum()
total_income_null_number = df['total_income'].isna().sum()

# вывод на экран с использованием f-строки
print(f'Пропусков в \'days_employed\' {days_employed_null_number} шт. ({(days_employed_null_number / len(df)):.0%})')
print(f'Пропусков в \'total_income\' {total_income_null_number} шт. ({(total_income_null_number / len(df)):.0%})')

Пропусков в 'days_employed' 2174 шт. (10%)
Пропусков в 'total_income' 2174 шт. (10%)


Количество пропущенных значений в обоих столбцах одинаково, отсюда можно предположить, что пропуски содержатся в одних и тех же строках таблицы. Проверим это с помощью логической индексации:

In [7]:
# создаем маску для условий индексации: ищем строки с пропусками в двух столбцах одновременно
mask = df['days_employed'].isna() & df['total_income'].isna()

# индексируем датафрейм по маске и смотрим его длину
len(df[mask])

2174

Длина датафрейма, содержащего пропуски одновременно в двух столбцах, равна количеству пропусков в каждом из столбцов.
Отсюда вывод: пропуски содержатся в двух столбцах одновременно в одних и тех же строках таблицы.

Информация о трудовом стаже клиента и размере его дохода является обязательной для получения кредита в банке, поэтому необходимо понять причины появления пропусков. Они могут быть такими:
* при загрузке/выгрузке данных произошел сбой (технические причины);
* при заполнении заявки на кредит сотрудником банка данные не внесены или внесены с ошибкой (человеческий фактор).

Если предположить, что в АБС банка отлажены механизмы последконтроля введенной сотрудником информации (обязательные поля не должны быть пустыми, в числовом поле не должно быть текстовых значений и пр.), то наиболее вероятная причина пропусков - технический сбой.



Чтобы не потерять 10% строк для дальнейшего анализа, необходимо заполнить пропуски данных в столбцах `days_employed` и `total_income`. Пропуски в количественных переменных заполняют характерными значениями, которые типичны для всей выборки.

Одно из возможных решений - заполнить пропуски медианным значением, т.е. числом, которое делит упорядоченную выборку пополам: ровно половина элементов больше медианы, вторая половина - меньше медианы. Применение медианы статистически не сильно исказит выборку (минимум, максимум, медиана останутся прежними) и, в отличие от среднего значения, отсечёт возможные выбросы (слишком большие или маленькие значения дохода/стажа). Применение среднего значения в данном случае менее оправданно, поскольку в датасете предложены разные категории клиентов: усреднять стаж и доходы пенсионеров и работающих нелогично.

Выведем описательную статистику данных в столбце с пропусками `days_employed`:

In [8]:
# получение статистической сводки о данных в столбце
df['days_employed'].describe().apply('{:.0f}'.format)

count     19351
mean      63046
std      140827
min      -18389
25%       -2747
50%       -1203
75%        -291
max      401755
Name: days_employed, dtype: object

<a id='intro1'></a>
В столбце `days_employed` с трудовым стажем данные слишком неоднородны: минимум выражен отрицательной величиной, медиана (50-й процентиль) тоже отрицательная. При этом трудовой стаж не может быть отрицательным, значит медианное значение искажено вследствие аномалий в данных. Кроме этого, максимальное значение трудового стажа составляет 401755 единиц, по описанию данных это должны быть дни, т.е. максимум будет 401755/365 = 1100 лет. Среднее значение также искажено и составляет 63046/365 = 173 года. По этому столбцу в следующем блоке проекта более детально исследуем аномалии и заполним пропуски после их устранения.

Проверим значения в столбце с пропусками `total_income`:

In [9]:
df['total_income'].describe().apply('{:.0f}'.format)

count      19351
mean      167422
std       102972
min        20667
25%       103053
50%       145018
75%       203435
max      2265604
Name: total_income, dtype: object

<a id='intro2'></a>
В столбце `total_income` с размером дохода данные соответствуют реальности: минимум 20 тыс.руб., максимум 2,3 млн.руб, среднее 167 тыс.руб., медиана 145 тыс.руб. Однако, по-моему мнению, заполнение пропусков общим медианным значением по столбцу - не самая хорошая идея, поскольку значения доходов у пенсионеров и работающих отличаются. Поэтому заполним пропуски медианным значением по каждой категории клиентов.

<a id='intro3'></a>
Для начала, создадим таблицу медианных значений доходов в разбивке по категориям клиентов:

In [10]:
df_grouped_1 = df.groupby('income_type')['total_income'].median()
df_grouped_1

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

Создадим функцию для одной строки, которая будет возвращать медианное значение дохода из сгруппированного датафрейма в том случае, если обнаружит пропуск в столбце с доходом. Если пропуска нет, то функция вернет исходное (заполненное) значение дохода:

In [11]:
def median_total_income(row): # подаем на вход строку датафрейма
    if pd.isna(row['total_income']): # если обнаружен (True) пропуск в столбце 'total_income', то
        return df_grouped_1.loc[row['income_type']] # вернем значение по соответствующему 'income_type' из вспомогат. таблицы
    return row['total_income'] # если пропуск не обнаружен, вернем текущее значение в 'total_income'

Перезапишем значение столбца с доходом с использованием метода .apply(), в который передадим созданную функцию и применим его к кажой строке таблицы:

In [12]:
df['total_income'] = df.apply(median_total_income, axis=1)

Если функция отработала верно, то медианное значение по всем категориям клиентов должно остаться прежним. Проверим это:

In [13]:
df.groupby('income_type')['total_income'].median() == df_grouped_1

income_type
безработный        True
в декрете          True
госслужащий        True
компаньон          True
пенсионер          True
предприниматель    True
сотрудник          True
студент            True
Name: total_income, dtype: bool

Кроме того, количество значений в перезаполненном столбце с заполненными пропусками должно совпасть с общим количеством строк в таблице:

In [14]:
df['total_income'].count() == len(df)

True

Посмотрим на итоговую статистику по столбцу с доходом во всей таблице по всем категориям клиентов:

In [15]:
df['total_income'].describe().apply('{:.0f}'.format)

count      21525
mean      165225
std        98044
min        20667
25%       107798
50%       142594
75%       195550
max      2265604
Name: total_income, dtype: object

Медианное значение изменилось незначительно: со 145 до 143 тыс.руб., однако пропуски теперь заполнены в соответствие с категорией клиента, что более логично.

**Выводы**

10% строк таблицы содержат пропуски данных одновременно в двух столбцах: трудовой стаж клиента и ежемесячный доход. Наиболее вероятными причинами пропусков могут являться технический сбой при загрузке/выгрузке данных или человеческий фактор.

Пропуски данных в столбце c размером дохода заполнены медианным значением в зависимости от типа занятости клиента, пропуски в столбце с трудовым стажем будут заполнены позднее после устранения аномалий в данных.

## Проверка данных на аномалии и исправления

Проверим и устраним аномалии в данных в столбцах с количественными переменными:

**данные о возрасте клиента в столбце `dob_years`:**

Требования банков по возрасту заемщиков - от 18 лет. Посмотрим на данные в этом столбце:

In [16]:
# проверим кол-во записей, где возраст менее 18 лет
df[df['dob_years'] < 18]['dob_years'].count()

101

In [17]:
# минимум и максимум в столбце
df['dob_years'].describe()[['min', 'max']]

min     0.0
max    75.0
Name: dob_years, dtype: float64

In [18]:
# проверим кол-во записей, где возраст 0 лет
df[df['dob_years'] == 0]['dob_years'].count()

101

В датасете встречается 101 строка, где возраст нулевой (в остальных строках возраст от 18 лет), посмотрим распределение строк с нулевым возрастом по сегментам клиентов (по типу занятости):

In [19]:
df[df['dob_years'] == 0].groupby(df['income_type'])['dob_years'].count()

income_type
госслужащий     6
компаньон      20
пенсионер      20
сотрудник      55
Name: dob_years, dtype: int64

Нулевые значения встречаются по всем сегментам клиентов: вероятно, произошла ошибка при вводе значения сотрудником банка или при загрузке/выгрузке данных.

Перезаписывать нулевые значения возраста будем на медианный возраст по сегментам клиентов. Алгоритм действий будет похож на [ранее](#intro3) использовавшийся:

In [20]:
# формируем дополнительную таблицу с медианным возрастом в каждом сегоменте клиентов
df_grouped_2 = df.groupby('income_type')['dob_years'].median()
df_grouped_2

income_type
безработный        38.0
в декрете          39.0
госслужащий        40.0
компаньон          39.0
пенсионер          60.0
предприниматель    42.5
сотрудник          39.0
студент            22.0
Name: dob_years, dtype: float64

In [21]:
# создаем функцию, возвращающую медианное значение возраста, если в исходнике возраст равен 0
def median_dob_year(row): # подаем на вход строку датафрейма
    if row['dob_years'] == 0: # если возраст равен 0, то
        return df_grouped_2.loc[row['income_type']] # вернем значение по соответствующему 'income_type' из вспомогат. таблицы
    return row['dob_years'] # если возраст ненулевой, то вернем текущее значение в 'dob_years'

In [22]:
# перезаписываем столбец с возрастом
df['dob_years'] = df.apply(median_dob_year, axis=1)

In [23]:
# проверка корректности работы функции: медиана не должна измениться по всем сегментам клиентов
df.groupby('income_type')['dob_years'].median() == df_grouped_2

income_type
безработный        True
в декрете          True
госслужащий        True
компаньон          True
пенсионер          True
предприниматель    True
сотрудник          True
студент            True
Name: dob_years, dtype: bool

Нулевые значения в столбце с возрастом клиента перезаписаны.

**данные о размере дохода в столбце `total_income`:**

Эти данные проверены [ранее](#intro2), пропуски заменены на медианные значения по сегментам клиентов, других аномалий не выявлено.

**данные о трудовом стаже клиента в столбце `days_employed`:**

[Ранее](#intro1) в этом столбце обнаружены отрицательные значения, попытаемся понять, почему они могли возникнуть.

Посчитаем количество положительных и отрицательных значений в столбце:

In [24]:
# кол-во строк с положительными значениями или нулем
df[df['days_employed'] >= 0]['days_employed'].count()

3445

In [25]:
# кол-во строк с отрицательными значениями
df[df['days_employed'] < 0]['days_employed'].count()

15906

Видим, что столбец заполнен преимущественно отрицательными значениями. Самое простое предположение: ошибки появились в результате сбоя при загрузке/выгрузке данных. Посмотрим детально: проверим заполнение положительными и отрицательными значениями в разбивке по типам занятости клиентов:

In [26]:
(df[df['days_employed'] >= 0]['days_employed'] # выбираем только строки с положительными значениями или нулем
    .groupby(df['income_type']) # группируем их по типам занятости
    .count() # считаем кол-во
)

income_type
безработный       2
пенсионер      3443
Name: days_employed, dtype: int64

In [27]:
df[df['days_employed'] < 0]['days_employed'].groupby(df['income_type']).count() # группировка для отрицательных значений стажа

income_type
в декрете              1
госслужащий         1312
компаньон           4577
предприниматель        1
сотрудник          10014
студент                1
Name: days_employed, dtype: int64

<a id='intro3'></a>
Предположим, что количество дней трудового стажа по пенсионерам (если этим маркером обозначаются именно неработающие пенсионеры) и безработным вносится в АБС банка вручную готовым значением, а для остальных типов ("занятые") рассчитывается в процессе оформления кредитной заявки от даты начала трудовой деятельности по состоянию "на сейчас". Тогда можно думать, что алгоритм расчета нарушен: отрицательное значение появляется в результате вычитания из даты начала трудовой деятельности даты подачи заявки на кредит, а не наоборот.

Устраним отрицательные значения в столбце `days_employed` путем замены всех значений на их модули:

In [28]:
# перезапись значений столбца с заменой на модули значений
df['days_employed'] = df['days_employed'].abs()

# контроль перезаписи значений: 
df[df['days_employed'] < 0]['days_employed'].count() == 0

True

Кроме того, [ранее](#intro1) выяснили, что необходимо разобраться со значениями в столбце, т.к. присутствуют аномально большие значения:

Посмотрим описательную статистику после перезаписи значений столбца:

In [29]:
df['days_employed'].describe().apply('{:.0f}'.format)

count     19351
mean      66915
std      139031
min          24
25%         927
50%        2194
75%        5538
max      401755
Name: days_employed, dtype: object

Видим, что медианное значение стало 2194 дня или около 6 лет. При этом остаются вопросы по максимальному и среднему значениям: они слишком велики. Посмотрим медианы в детальной разбивке по типам занятости клиентов:

In [30]:
df.groupby('income_type')['days_employed'].median().apply('{:.0f}'.format)

income_type
безработный        366414
в декрете            3297
госслужащий          2689
компаньон            1547
пенсионер          365213
предприниматель       521
сотрудник            1574
студент               579
Name: days_employed, dtype: object

Становится понятным, что сегменты клиентов "пенсионер" и "безработный" снова выбиваются из общей массы аномальными значениями, как было и [ранее](#intro3). Не исключены как влияние человеческого фактора (например, кредитный сотрудник мог внести данные не в днях, а в часах), так и технический сбой.

Предположение о том, что данные приведены в часах, на первый взгляд дает хорошую картинку: медианный стаж становится равен 365213/24/365 = 42 года. При этом логика сбивается на сегменте "безработные":

In [31]:
# для удобства анализа добавим столбец, отражающий стаж в годах
df.insert(2, 'years_employed', df['days_employed']/24/365)

In [32]:
df[df['income_type'] == 'безработный']

,children,days_employed,years_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
3133,1,337524.466835,38.530190,31.0,среднее,1,женат / замужем,0,M,безработный,1,59956.991984,покупка жилья для сдачи
14798,0,395302.838654,45.125895,45.0,Высшее,0,гражданский брак,1,F,безработный,0,202722.511368,ремонт жилью


Возраст клиентов категории "безработные" 31 год и 45 лет, при этом если стаж выражен в часах, то он получается 39 и 45 лет соответственно: первый начал работать до рождения, второй работает с рождения.

По категории "пенсионер" видим аналогичную ситуацию: в таблице есть 52 строки, где трудовой стаж больше возраста клиента:

In [33]:
len(df[(df['income_type'] == 'пенсионер') & (df['years_employed'] > df['dob_years'])])

52

Отвергаем гипотезу о том, что по пенсионерам и безработным столбец со стажем заполнен не количеством дней, а количеством часов.

Дополнительно проверим, есть ли в таблице еще записи, где трудовой стаж больше возраста клиента (исключим из рассмотрения пенсионеров и безработных):

In [34]:
len(
df[~(df['income_type'] == 'пенсионер') &
   ~(df['income_type'] == 'безработный') &
    (df['years_employed'] > df['dob_years'])]
)

0

В остальных строках описанные расхождения отсутствуют.

При отсутствии возможности задать дополнительные вопросы единственное решение, которое пришло на ум - перезаполнить стаж пенсионеров и безработных максимально возможным значением по формуле (возраст - 18 лет). Можно взять и 16 лет (минимальный возраст трудоустройства), но выбираю 18 лет исключительно из анализа собственного окружения.

In [35]:
# создаем функцию, возвращающую значение максимально возможного стажа для "проблемных" категорий клиентов
def refresh_days_employed(row): # подаем на вход строку датафрейма
    if row['income_type'] in ['пенсионер', 'безработный']: # если категория клиента 'пенсионер' или 'безработный'
        return (row['dob_years'] - 18)*365  # рассчитаем и вернем макс.возможное значение стажа
    return row['days_employed'] # для остальных категорий возвращаем текущее значение стажа

In [36]:
# перезаписываем столбец со стажем
df['days_employed'] = df.apply(refresh_days_employed, axis=1)

In [37]:
# проверка перезаписи значений: в итоговом df не должно быть расхождений возраста и стажа
df['years_employed'] = df['days_employed']/24/365
len(df[df['years_employed'] > df['dob_years']])

0

In [38]:
# удаляем созданный для анализа столбец со стажем в годах
df = df.drop(columns='years_employed')

Посмотрим описательную статистику после перезаписи значений столбца:

In [39]:
df['days_employed'].describe().apply('{:.0f}'.format)

count    19764
mean      4841
std       5553
min         24
25%        944
50%       2269
75%       6177
max      20440
Name: days_employed, dtype: object

Сейчас данные отражают реальность: минимальный, максимальный, средний и медианный стаж работы составляют соответственно 24 дня, 56 лет, 13 лет и 5 лет.

Осталось заполнить пропущенные значения: снова будем идти от медианного значения по каждой категории клиентов.

In [40]:
# формируем дополнительную таблицу с медианным стажем в каждом сегменте клиентов
df_grouped_3 = df.groupby('income_type')['days_employed'].median()
df_grouped_3

income_type
безработный         7300.000000
в декрете           3296.759962
госслужащий         2689.368353
компаньон           1547.382223
пенсионер          15330.000000
предприниматель      520.848083
сотрудник           1574.202821
студент              578.751554
Name: days_employed, dtype: float64

In [41]:
# создаем функцию, возвращающую медианное значение стажа для пропущенных значений
def median_days_employed(row): # подаем на вход строку датафрейма
    if pd.isna(row['days_employed']): # если обнаружен (True) пропуск в столбце 'days_employed', то
        return df_grouped_3.loc[row['income_type']] # вернем значение по соответствующему 'income_type' из вспомогат. таблицы
    return row['days_employed'] # если пропуск не обнаружен, вернем текущее значение в 'days_employed'

In [42]:
# перезаписываем столбец со стажем
df['days_employed'] = df.apply(median_days_employed, axis=1)

In [43]:
# проверка корректности работы функции: медианное значение должно сохраниться по всем сегментам клиентов
df.groupby('income_type')['days_employed'].median() == df_grouped_3

income_type
безработный        True
в декрете          True
госслужащий        True
компаньон          True
пенсионер          True
предприниматель    True
сотрудник          True
студент            True
Name: days_employed, dtype: bool

In [44]:
# проверка количества значений в перезаписанном столбце на соответствие общей длине таблицы
df['days_employed'].count() == len(df)

True

In [45]:
# проверка пропущенных значений в таблице после всех преобразований:
df.isna().sum().sum() == 0

True

В данных столбца с трудовым стажем устранены аномалии, пропущенные значения заполнены медианным значением стажа в зависимости от сегомента клиента.

**данные о количестве детей в столбце `children`:**

In [46]:
# посмотрим на уникальные значения и их кол-во в датасете
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Очевидно, что значения -1 и 20 для этого набора данных являются аномальными: произошла ошибка при загрузке/выгрузке или вводе данных при заполнении кредитной заявки в АБС банка. Если ошибка возникла под воздействием человеческого фактора, то можно предположить, что на месте этих значений должны быть 1 и 2 соответственно. Однако, перезаписать эти данные на предполагаемое или какое-либо характерное значение (среднее, мода, медиана), на мой взгляд, не будет хорошей идеей, т.к. может привести к искажению итоговых расчетов, ведь показатель количества детей фигурирует в проверке главной гипотезы.

Поскольку аномальных значений всего (76 + 47) = 123, что составляет 123 / 21525 = 0,6% общего кол-ва значений в столбце, то удалим эти строки из датасета и проверять гипотезу будем на оставшихся строках:

In [47]:
# отбор строк с аномальными значениями и исключение их из df со сбросом старых индексов и перезаписью таблицы
df = df[~(df['children'] == -1) & ~(df['children'] == 20)].reset_index(drop=True)

# проверка количества удаленных строк
21525 - len(df) == 123

True

**Выводы**

Произведены изменения исходного датасета:

* обнаружены и устранены аномалии;
* заполнены ячейки с пропущенными данными.

## Изменение типов данных

Смотрим информацию об измененной таблице:

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21402 entries, 0 to 21401
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21402 non-null  int64  
 1   days_employed     21402 non-null  float64
 2   dob_years         21402 non-null  float64
 3   education         21402 non-null  object 
 4   education_id      21402 non-null  int64  
 5   family_status     21402 non-null  object 
 6   family_status_id  21402 non-null  int64  
 7   gender            21402 non-null  object 
 8   income_type       21402 non-null  object 
 9   debt              21402 non-null  int64  
 10  total_income      21402 non-null  float64
 11  purpose           21402 non-null  object 
dtypes: float64(3), int64(4), object(5)
memory usage: 2.0+ MB


Как видим, все пропуски данных устранены, при этом тип данных с столбце с возрастом клиента `dob_year` изменился на вещественный. Это произошло вследствие заполнения пропусков в столбце медианным значением, которое не обязано быть целочисленным.

Изменим вещественный тип данных на целочисленный:

In [49]:
# изменение типа данных с перезаписью значений столбца
df['days_employed'] = df['days_employed'].astype('int')

In [50]:
df['dob_years'] = df['dob_years'].astype('int')

In [51]:
df['total_income'] = df['total_income'].astype('int')

In [52]:
# проверка типов данных после замены:
df[['days_employed', 'dob_years', 'total_income']] .info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21402 entries, 0 to 21401
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   days_employed  21402 non-null  int32
 1   dob_years      21402 non-null  int32
 2   total_income   21402 non-null  int32
dtypes: int32(3)
memory usage: 250.9 KB


**Выводы**

В отдельных столбцах таблицы переменные переведены из вещественного типа данных в целочисленный. Это уменьшит объем хранимой информации, упростит визуальное восприятие данных и повлияет на скорость их обработки.

## Удаление дубликатов

Проверим датасет на наличие неявных дубликатов в категориальных переменных c помощью метода .value_counts(). Чаще всего, выбираю этот метод, т.к., в отличие от метода .unique(), в результате видны не только уникальные имена переменных, но и их количество. Однако, результаты вывода метода .unique() посредством массива визуально более компактны (если уникальных значений много) и более удобны для копирования при дальнейшей обработке/замене данных.

**данные в столбце `education`**

In [53]:
# вывод уникальных значений столбца
df['education'].value_counts()

среднее                13667
высшее                  4698
СРЕДНЕЕ                  766
Среднее                  703
неоконченное высшее      665
ВЫСШЕЕ                   271
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Неявные дубликаты в этом столбце появились вследствие разного регистра в строковых данных. Причина их появления, скорее всего, - в АБС банка не автоматизирован ввод и последконтроль данных об образовании (вносятся вручную); исправим это:

In [54]:
# приведение символов в строковых значениях столбца к нижнему регистру с перезаписью столбца
df['education'] = df['education'].str.lower()

In [55]:
# контроль устранения неявных дубликатов
df['education'].value_counts()

среднее                15136
высшее                  5237
неоконченное высшее      741
начальное                282
ученая степень             6
Name: education, dtype: int64

**данные в столбце `education_id`**

In [56]:
# вывод уникальных значений столбца
df['education_id'].value_counts()

1    15136
0     5237
2      741
3      282
4        6
Name: education_id, dtype: int64

Эта переменная по сути является категориальной и представляет из себя идентификатор уровня образования, выраженный целым числом. Ранжировать и сравнивать данные по значению этой переменной нельзя.

Неявных дубликатов в столбце не обнаружено.

**данные в столбце `family_status`**

In [57]:
# вывод уникальных значений столбца
df['family_status'].value_counts()

женат / замужем          12302
гражданский брак          4160
Не женат / не замужем     2799
в разводе                 1189
вдовец / вдова             952
Name: family_status, dtype: int64

Неявных дубликатов в столбце не обнаружено. Приведем данные к нижнему регистру:

In [58]:
# приведение символов в строковых значениях столбца к нижнему регистру с перезаписью столбца
df['family_status'] = df['family_status'].str.lower()

In [59]:
# контроль перезаписи
df['family_status'].value_counts()

женат / замужем          12302
гражданский брак          4160
не женат / не замужем     2799
в разводе                 1189
вдовец / вдова             952
Name: family_status, dtype: int64

**данные в столбце `family_status_id`**

In [60]:
# вывод уникальных значений столбца
df['family_status_id'].value_counts()

0    12302
1     4160
4     2799
3     1189
2      952
Name: family_status_id, dtype: int64

Эта переменная по сути является категориальной и представляет из себя идентификатор семейного статуса, выраженный целым числом. Ранжировать и сравнивать данные по значению этой переменной нельзя.

Неявных дубликатов в столбце не обнаружено.

**данные в столбце `gender`**

In [61]:
# вывод уникальных значений столбца
df['gender'].value_counts()

F      14154
M       7247
XNA        1
Name: gender, dtype: int64

В датасете присутствует клиент, пол которого не определен: скорее всего, это неявный пропуск с нестандартным обозначением пропущенного значения (хотя можно предположить, что символами XNA обозначается один из нескольких десятков вариантов гендерной идентичности). Поддавшись влиянию гендерных стереотипов и учитывая то, что удаление всего одной записи не повлияет на чистоту эксперимента, удалим эту строку из датасета:

In [62]:
# отсекаем лишнюю строку и перезаписываем df со сбросом старых индексов
df = df[df['gender'] != 'XNA'].reset_index(drop=True)

In [63]:
# контроль
len(df) == 21525-123-1

True

Из датасета успешно удалена строка с неявным пропуском в категориальной переменной.

**данные в столбце `income_type`**

In [64]:
# вывод уникальных значений столбца
df['income_type'].value_counts()

сотрудник          11050
компаньон           5053
пенсионер           3839
госслужащий         1453
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64

Неявных дубликатов в столбце не обнаружено.

**данные в столбце `debt`**

In [65]:
df['debt'].value_counts()

0    19669
1     1732
Name: debt, dtype: int64

Неявных дубликатов в столбце не обнаружено. 8% клиентов выходили на просрочку платежей по кредиту.

**данные в столбце `purpose`**

In [66]:
# вывод уникальных значений столбца
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Данные в этом столбце, скорее всего, вносятся в кредитную заявку "со слов" клиента, поэтому они такие разнообразные. Хотя, по сути, целей кредита здесь всего 4: автомобиль, недвижимость, свадьба, образование. Категоризацию данных выполним позднее.

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма

Создадим новые датафреймы-словари для оптимизации объема исходной таблицы, упрощения работы с данными (фильтрацию по числовой метке реализовать удобнее, чем вносить текст) и визуального восприятия таблицы.

In [67]:
# отбираем нужные столбцы, удаляем дубликаты и сбрасываем индексы; записываем в новый датафрейм
education_dict = df[['education', 'education_id']].drop_duplicates().reset_index(drop=True)

In [68]:
# контроль результата
education_dict

,education,education_id
0,высшее,0
1,среднее,1
2,неоконченное высшее,2
3,начальное,3
4,ученая степень,4


In [69]:
family_status_dict = df[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)

In [70]:
family_status_dict

,family_status,family_status_id
0,женат / замужем,0
1,гражданский брак,1
2,вдовец / вдова,2
3,в разводе,3
4,не женат / не замужем,4


Удалим из исходного датафрейма столбцы `education` и `family_status`:

In [71]:
# удаление колонок с использованием их названий
df = df.drop(columns=['education', 'family_status'])

In [72]:
# контроль обновленной таблицы
df.head()

,children,days_employed,dob_years,education_id,family_status_id,gender,income_type,debt,total_income,purpose
0,1,8437,42,0,0,F,сотрудник,0,253875,покупка жилья
1,1,4024,36,1,0,F,сотрудник,0,112080,приобретение автомобиля
2,0,5623,33,1,0,M,сотрудник,0,145885,покупка жилья
3,3,4124,32,1,0,M,сотрудник,0,267628,дополнительное образование
4,0,12775,53,1,1,F,пенсионер,0,158616,сыграть свадьбу


**Выводы**

Датафремы-словари для классификации уровней образования и семейного положения созданы, исходная таблица оптимизирована: удалены соответствующие столбцы со строковыми обозначениями, оставлены только их идентификаторы.

## Категоризация дохода

Присвоим категорию доходу каждого клиента в зависимости от размера дохода:

* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

Задачу выполним через наполнение списка с помощью цикла for, потенциальные ошибки в работе цикла обработаем с помощью конструкции try...except..., полученный список трансформируем в колонку и добавим ее в датафрейм. 

*Примечание:* Первоначально я хотела использовать функцию одной строки и добавление столбца в таблицу через применение к колонке с доходом метода .apply() с переданной в него функцией. Однако в ТЗ по проекту описанный способ предлагается применить при категоризации данных о цели кредита, в связи с чем решила попробовать альтернативный способ. Возможно, он приводит к более медленному расчету результата, но он работает.

Создадим в таблице новый столбец, отражающий категорию клиента по доходу.

In [73]:
category_list = [] # создаем пустой список

for i in range(len(df)): # создаем цикл с перебором индексов от 0 до <длина датафрейма>
    try:
        # учим цикл перебирать строки таблицы,
        if 0 <= df.loc[i, 'total_income'] <=30000: # сравнивать значение в колонке 'total_income' с диапазонами
            category = 'E' # и сохранять в переменную category обозначение категории 
        elif 30001 <= df.loc[i, 'total_income'] <=50000:
            category = 'D'
        elif 50001 <= df.loc[i, 'total_income'] <=200000:
            category = 'C'
        elif 200001 <= df.loc[i, 'total_income'] <=1000000:
            category = 'B'
        elif df.loc[i, 'total_income'] >= 1000001:
            category = 'A'
        category_list.append(category) # добавляем обозначение категории в конец списка
    except:
        category_list.append('ошибка: категория не присвоена') # значение при ошибке (отрицательный доход)

В результате получили список категорий, сформированный в порядке возрастания индексов df. Добавим значения из этого списка в отдельную колонку.
Новую колонку можно создать методом pd.Series, но его результат меня не устроил, т.к. новая колонка добавляется в конец df. Применила метод .insert() с указанием позиции для вставки нового столбца

In [74]:
# добавляем новый столбец после столбца 9, метод .insert() перезаписывает исходный df, присваивание можно опустить
df.insert(9, 'total_income_category', category_list)

In [75]:
# визуальный контроль результата
df.head()

,children,days_employed,dob_years,education_id,family_status_id,gender,income_type,debt,total_income,total_income_category,purpose
0,1,8437,42,0,0,F,сотрудник,0,253875,B,покупка жилья
1,1,4024,36,1,0,F,сотрудник,0,112080,C,приобретение автомобиля
2,0,5623,33,1,0,M,сотрудник,0,145885,C,покупка жилья
3,3,4124,32,1,0,M,сотрудник,0,267628,B,дополнительное образование
4,0,12775,53,1,1,F,пенсионер,0,158616,C,сыграть свадьбу


In [76]:
# контроль результата: проверка категорий условиям категоризации
df.pivot_table(index='total_income_category', values = 'total_income', aggfunc = ['min', 'median', 'max'])

,min,median,max
,total_income,total_income,total_income
total_income_category,,,
A,1004476,1223042.0,2265604
B,200001,257619.0,997014
C,50025,130426.5,199980
D,30075,43929.0,49996
E,20667,26431.0,29749


**Вывод**

Произведена категоризация доходов клиентов, новый столбец с категорией дохода добавлен в таблицу. 

## Категоризация целей кредита

Необходимо создать новый столбец таблицы, который определит категории для целей кредита из предложенного списка:
* 'операции с автомобилем',
* 'операции с недвижимостью',
* 'проведение свадьбы',
* 'получение образования'

Поставленную задачу предлагается решить посредством создания собственной функции и передачи ее в метод .apply() для добавление нового столбца в датафрейм.

В столбце `purpose` содержатся следующие значения:

In [77]:
# вывод уникальных значений столбца
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [78]:
def find_str(text): # задаем функцию, принимающую на вход некий text
    if text.find('авто') >= 0: # ищем неотрицательный индекс вхождения подстроки 'авто' в переданный text и т.д.
        return 'операции с автомобилем'
    if text.find('недвиж') >= 0 or text.find('жиль') >= 0:
        return 'операции с недвижимостью'
    if text.find('свадьб') >= 0:
        return 'проведение свадьбы'
    if text.find('образов') >= 0:
        return 'получение образования'
    return 'ошибка: категория не присвоена' # если ни один из вариантов подстрок не подойдет, то вернется ошибка

In [79]:
# создаем новый столбец датафрейма на базе существующего столбца
df['purpose_category'] = df['purpose'].apply(find_str)

In [80]:
# визуальный контроль результата
df.head()

,children,days_employed,dob_years,education_id,family_status_id,gender,income_type,debt,total_income,total_income_category,purpose,purpose_category
0,1,8437,42,0,0,F,сотрудник,0,253875,B,покупка жилья,операции с недвижимостью
1,1,4024,36,1,0,F,сотрудник,0,112080,C,приобретение автомобиля,операции с автомобилем
2,0,5623,33,1,0,M,сотрудник,0,145885,C,покупка жилья,операции с недвижимостью
3,3,4124,32,1,0,M,сотрудник,0,267628,B,дополнительное образование,получение образования
4,0,12775,53,1,1,F,пенсионер,0,158616,C,сыграть свадьбу,проведение свадьбы


In [81]:
# еще один контроль результата: все ли строки категоризированы
len(df[df['purpose_category'] == 'ошибка: категория не присвоена'])

0

In [82]:
# проверка уникальных значений в новом столбце
df['purpose_category'].value_counts()

операции с недвижимостью    10779
операции с автомобилем       4288
получение образования        3997
проведение свадьбы           2337
Name: purpose_category, dtype: int64

Значение 'ошибка: категория не присвоена' в датафрейме отсутствует, значит поисковые подстроки в функцию для категоризации переданы верно.

**Вывод**

Произведена категоризация целей получения кредита, новый столбец с категорией добавлен в таблицу. 

## Поиск явных дубликатов, итоговая очистка датафрейма

Проверим датасет на наличие явных дубликатов и удалим их при обнаружении:

In [83]:
# подсчет количества дубликатов
df.duplicated().sum()

71

In [84]:
# удаление дубликатов со сбросом старых индексов и перезаписью датафрейма
df = df.drop_duplicates().reset_index(drop=True)

In [86]:
# контроль
len(df) == 21525-123-1-71

True

Явные дубликаты обнаружены и удалены.

## Влияние различных клиентских характеристик на возврат кредита в срок

**Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Проверять зависимость будем через вывод сводной таблицы, строками которой будет количество детей, столбцами - факт нарушения графика возврата, а данные будут отображать количество записей в разных комбинациях "дети - факт нарушения срока возврата" (считать количество записей можно по любому столбцу таблицы, т.к. все пропуски устранены). В эту таблицу внедряем столбец, отражающий долю клиентов, нарушивших срок возврата кредита, в общем количестве клиентов каждой группы.

In [87]:
# строим сводную таблицу
df_pivot_children = df.pivot_table(index='children', columns='debt', values='dob_years', aggfunc='count', fill_value=0) \
                      .rename(columns={0: 'payment_on_time', 1: 'delay_in_payment'}) # даем столбцам понятные наименования

# добавляем столбец с расчетом доли
df_pivot_children['delay_in_payment_%'] = (df_pivot_children['delay_in_payment'] / \
                                          (df_pivot_children['payment_on_time'] + df_pivot_children['delay_in_payment']) \
                                          * 100) \
                                          .round(1)
    
df_pivot_children

debt,payment_on_time,delay_in_payment,delay_in_payment_%
children,,,
0,13027,1063,7.5
1,4364,444,9.2
2,1858,194,9.5
3,303,27,8.2
4,37,4,9.8
5,9,0,0.0


**Вывод: Наблюдаем выраженную зависимость между количеством детей и возвратом кредита в срок**:
* минимальный процент клиентов, нарушивших срок возврата кредита, наблюдается в группе клиентов, не имеющих детей, и составляет 7,5%;
* в группе клиентов с 1-3 детьми доля клиентов, нарушивших срок возврата кредитов, повышается и составляет от 8,2% до 9,5%. Самое дисциплинированное поведение демонстрируют клиенты с 3 детьми, что может быть связано с более осознанным подходом к планированию бюджета семьи;
* максимальный процент клиентов, нарушивших срок возврата кредита, наблюдается в группе клиентов с 4-мя детьми, и составляет 9,8%: форс-мажорные траты в таких семьях могут возникать чаще, чем в остальных;
* у клиентов с 5 детьми фактов нарушения графика возврата кредита не фиксируется, что может быть связано как с тем, что такие семьи очень трезво оценивают возможности по оплате кредита, так и с недостаточностью объема выборки для объективного вывода (если хотя бы 1 из 9 таких клиентов нарушит график возврата, то это приведет бы к расчету доли в 11% и также подтвердит гипотезу о зависимости количества детей и возврата кредита в срок).

**Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Проверять зависимость будем по аналогии с предыдущим выводом, дополнительно объединим сформированную сводную таблицу с таблицей-словарем для улучшения визуального восприятия.

In [88]:
# строим сводную таблицу
df_pivot_family_status = df.pivot_table(index='family_status_id', columns='debt', values='dob_years', aggfunc='count') \
                      .rename(columns={0: 'payment_on_time', 1: 'delay_in_payment'}) # даем столбцам понятные наименования

# добавляем столбец с расчетом доли
df_pivot_family_status['delay_in_payment_%'] = \
            (df_pivot_family_status['delay_in_payment'] / \
            (df_pivot_family_status['payment_on_time'] + df_pivot_family_status['delay_in_payment']) \
            *100) \
            .round(1)

# объединяем со таблицей-словарем для отображения семейного статуса c удалением неинформативного столбца
family_status_dict.merge(df_pivot_family_status, on='family_status_id', how='left') \
                  .drop(columns='family_status_id') \
                  .sort_values(by='delay_in_payment_%')

,family_status,payment_on_time,delay_in_payment,delay_in_payment_%
2,вдовец / вдова,888,63,6.6
3,в разводе,1105,84,7.1
0,женат / замужем,11334,927,7.6
1,гражданский брак,3748,385,9.3
4,не женат / не замужем,2523,273,9.8


Дополнительно выведем сводную таблицу, отражающую медианный возраст клиентов - она поможет в формировании выводов:

In [89]:
family_status_dict.merge(df.pivot_table(index='family_status_id', columns='debt', values='dob_years', aggfunc='median'), \
                         on='family_status_id') \
                        .rename(columns={0: 'payment_on_time_med_age', 1: 'delay_in_payment_med_age'}) \
                        .drop(columns='family_status_id')

,family_status,payment_on_time_med_age,delay_in_payment_med_age
0,женат / замужем,43.0,39.0
1,гражданский брак,41.5,38.0
2,вдовец / вдова,58.0,57.0
3,в разводе,46.0,44.0
4,не женат / не замужем,36.0,32.0


**Вывод: Наблюдаем выраженную зависимость между семейным положением и возвратом кредита в срок**:
* минимальный процент клиентов, нарушивших срок возврата кредита, наблюдается в группе клиентов "вдовец / вдова", и составляет 6,6%. Это может быть связано с тем, что эти клиенты более возрастные (57 года), их траты на бытовые нужды семьи снижены в связи с отсутствием супруга и малолетних детей;
* в группе клиентов "в разводе" доля клиентов, нарушивших срок возврата кредитов, повышается и составляет 7,1%. Это может быть связано с тем, что по сравнению с первой группой эти клиенты более молодые (44 года), и в их траты входят, скорее всего, алименты на содержание детей от предыдущих браков;
* в самой многочисленной группе клиентов "женат / замужем" доля клиентов, нарушивших срок возврата кредитов, повышается и составляет 7,6%. Это незначительное повышение по сравнению с предыдущей группой может быть связано с возникновением дополнительных незапланированных трат на нужды семьи;
* значительное повышение доли (по сравнению с предыдущими группами клиентов) наблюдается у клиентов группы "гражданский брак": доля клиентов-"просрочников" составляет уже 9,3%. Средний возраст таких клиентов в представленном датасете критично не отличается от клиентов, зарегистрировавших брак официально. Опять поддаюсь влиянию стереотипов и предполагаю, что это более легкомысленные клиенты, не стремящиеся к четкому исполнению обязательств (как перед семьей в традиционном ее понимании, так и перед банком);
* максимальный процент клиентов, нарушивших срок возврата кредита, наблюдается в группе неженатых / незамужних клиентов: 9,8%. На это, вероятнее всего, влияет относительно молодой возраст таких клиентов (32 года) и недостаточные навыки планирования собственного бюджета.

P.S. Очевидно, что семейное положение и возраст клиента неразрывно связаны: общество пока живет по традиционным устоям, хотя возрастной порог вхождения в официальный брак постепенно увеличивается. Тем не менее, пока можно говорить о том, что именно возраст клиента вкупе с его семейным положением влияют на исполнение обязательств по кредиту (жаль, что навык планирования бюджета не является врожденным).

**Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Проверять зависимость будем по аналогии с предыдущими выводами, дополнительно создадим таблицу-словарь для улучшения визуального восприятия результирующей таблицы.

In [90]:
# дополнительный датафрейм-словарь: каждой категории соответствует свой диапазон доходов
data = [['A', '1000001 and more'],
        ['B', '200001–1000000'],
        ['C', '50001–200000'],
        ['D', '30001–50000'],
        ['E', '0–30000']]
columns=['total_income_category', 'total_income_range']

category_dict = pd.DataFrame(data=data, columns=columns)

In [91]:
# строим сводную таблицу
df_pivot_total_income = df.pivot_table(index='total_income_category', columns='debt', values='dob_years', aggfunc='count') \
                      .rename(columns={0: 'payment_on_time', 1: 'delay_in_payment'})

# добавляем столбец с расчетом доли
df_pivot_total_income['delay_in_payment_%'] = \
            (df_pivot_total_income['delay_in_payment'] / \
            (df_pivot_total_income['payment_on_time'] + df_pivot_total_income['delay_in_payment']) \
            *100) \
            .round(1)

category_dict.merge(df_pivot_total_income, on='total_income_category')

,total_income_category,total_income_range,payment_on_time,delay_in_payment,delay_in_payment_%
0,A,1000001 and more,23,2,8.0
1,B,200001–1000000,4659,354,7.1
2,C,50001–200000,14568,1353,8.5
3,D,30001–50000,328,21,6.0
4,E,0–30000,20,2,9.1


**Вывод: Наблюдаем нелинейную зависимость между уровнем дохода и возвратом кредита в срок:**

* самыми дисциплинированными с точки зрения возврата кредита в срок являются клиенты категории D с доходом от 30 до 50 тыс.руб.: среди них доля "просрочников" составляет 6%. Можно предположить, что эти клиенты, имея относительно небольшой доход, обладают самыми развитыми навыками планирования бюджета, в т.ч. они могут иметь неофициальную подработку;
* если смотреть только на показатели самых высоко- и низкодоходных клиентов (категории А и Е), то видим доли "просрочников" 8% и 9,1% соответственно, причем самые низкодоходные клиенты (доход до 30 тыс.руб.) возглавляют список проблемных клиентов по сравнению со всеми остальными категориями. Однако при движении между категориями А и Е не наблюдается линейной зависимости между уровнем дохода и возвратом кредита в срок;
* из трех самых многочисленных групп клиентов наихудшее поведение с точки зрения возврата кредита в срок демонстрируют клиенты категории С с доходом от 50 до 200 тыс.руб. Возможно, это связано с тем, что такие клиенты имеют единственный источник дохода по основному месту работы, и любые финансовые трудности (задержка выплаты зарплаты, форс-мажор в семье) приводят к сложностям.

**Как разные цели кредита влияют на его возврат в срок?**

Проверять зависимость будем по аналогии с предыдущими выводами через сводную таблицу.

In [92]:
# строим сводную таблицу
df_pivot_purpose = df.pivot_table(index='purpose_category', columns='debt', values='dob_years', aggfunc='count') \
                      .rename(columns={0: 'payment_on_time', 1: 'delay_in_payment'})

# добавляем столбец с расчетом доли
df_pivot_purpose['delay_in_payment_%'] = (df_pivot_purpose['delay_in_payment'] / \
                                         (df_pivot_purpose['payment_on_time'] + df_pivot_purpose['delay_in_payment']) \
                                         *100) \
                                         .round(1)

df_pivot_purpose.sort_values(by='delay_in_payment_%')

debt,payment_on_time,delay_in_payment,delay_in_payment_%
purpose_category,,,
операции с недвижимостью,9970,780,7.3
проведение свадьбы,2130,183,7.9
операции с автомобилем,3879,400,9.3
получение образования,3619,369,9.3


**Вывод: Наблюдаем выраженную зависимость между целью кредита и возвратом кредита в срок:**
* наименьший процент клиентов, имеющих нарушения сроков возврата кредита, фиксируется по кредитам на недвижимость и составляет 7,3%: это наиболее серьезная с точки зрения финансовых вложений сделка для любого клиента, поэтому к ее планированию подходят более осознанно;
* наибольший процент клиентов-"просрочников" наблюдается по целям кредитов на автомобиль и образование. Вероятно, это связано с более короткими сроками кредитов на эти цели и значительными суммами ежемесячных платежей;
* по кредитам на проведение свадьбы, честно, мне ничего в голову не приходит. Эти клиенты просто в общем тренде: по всему датасету доля "просрочников" 8%.

## Общий вывод:

Для анализа платежеспособности клиентов и ответов на вопросы исследования получена таблица с данными по клиентам банка, после изучения которой проведены следующие действия:
* заполнены пропущенные значения, устранены аномалии, удалены полные дубликаты строк и исправлены неявные дубликаты: это помогает получить "чистые" данные для дальнейших исследований;
* изменены типы данных, созданы таблицы-словари, удалены отдельные столбцы: это помогает оптимизировать объем информации, увеличить скорость обработки данных и улучшает визуальное восприятие таблицы;
* произведена категоризация клиентов по уровню доходов и целям кредитов: это дает возможности для анализа данных обобщенно по группам клиентов.

В ходе проведенного исследования установлены следующие факты:
* наблюдается выраженная зависимость между количеством детей и возвратом кредита в срок: клиенты с детьми потенциально чаще могут нарушать сроки возврата кредита по сравнению с клиентами без детей;
* наблюдается выраженная зависимость между семейным положением и возвратом кредита в срок: снижение доли клиентов, нарушающих срок возврата кредита, происходит линейно между статусами от "неженат" до "вдовец", что в большей степени обусловлено возрастом клиентов и развитием навыков планирования бюджета;
* наблюдается нелинейная зависимость между уровнем дохода и возвратом кредита в срок: наиболее осознанно к возврату кредита подходят клиенты с относительно небольшим (но не минимальным) уровнем дохода;
* наблюдается выраженная зависимость между целью кредита и возвратом кредита в срок: клиенты, оформляющие кредиты на короткие сроки со значительными суммами ежемесячных платежей (например, автокредиты) потенциально чаще нарушают требования по возврату кредита по сравнению с клиентами, оформляющими кредит для операций с недвижимостью.

Вышеописанные тезисы могут быть учтены при построении/доработке модели кредитного скоринга.